plugins/woocommerce/src/Blocks/Domain/Services/CheckoutLink.php

```diff
@@ -94,21 +94,70 @@ protected function get_products_from_checkout_link() {
 		$raw_products = array_filter( explode( ',', wc_clean( wp_unslash( $_GET['products'] ?? '' ) ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
 		$products     = [];
 
-		foreach ( $raw_products as $product_id_qty ) {
-			if ( strpos( $product_id_qty, ':' ) !== false ) {
-				list( $product_id, $qty ) = explode( ':', $product_id_qty );
+		foreach ( $raw_products as $raw_product ) {
+			$cart_item_data = [];
+
+			if ( strpos( $raw_product, ':' ) !== false ) {
+
+				list( $product_id, $qty, $product_data ) = array_pad( explode( ':', $raw_product, 3 ), 3, false );
+
+				// If the quantity is not-numeric, it may be key=value pairs.
+				if ( ! is_numeric( $qty ) ) {
+					$product_data = $qty;
+					$qty          = 1;
+				}
+
+				if ( $product_data ) {
+
+					// Parse key=value pairs separated by semicolons
+					$pairs = explode( ';', $product_data );
+
+					$cart_item_data = [];
+
+					foreach ( $pairs as $pair ) {
+						if ( strpos( $pair, '=' ) === false ) {
+							continue;
+						}
+						list( $key, $value ) = explode( '=', $pair, 2 );
+						$key                 = trim( $key );
+						$value               = trim( $value );
+
+						$cart_item_data[ $key ] = wc_clean( $value );
+					}
+				}
 			} else {
-				$product_id = $product_id_qty;
+				$product_id = $raw_product;
 				$qty        = 1;
 			}
+
+			if ( ! is_numeric( $product_id ) ) {
+				// If the product ID is not numeric, it may be a SKU.
+				$product_id = wc_get_product_id_by_sku( wc_clean( $product_id ) );
+			}
+
 			$product_id = absint( $product_id );
 			$qty        = absint( $qty );
 
 			if ( ! $product_id || ! $qty ) {
 				continue;
 			}
 
-			$products[ $product_id ] = $qty;
+			$variation = array_map(
+				function ( $key, $value ) {
+					return [
+						'attribute' => $key,
+						'value'     => $value,
+					];
+				},
+				array_keys( $cart_item_data ),
+				$cart_item_data
+			);
+
+			$products[ $product_id ] = [
+				'quantity'       => $qty,
+				'variation'      => $variation,
+				'cart_item_data' => $cart_item_data,
```

Comment:
Is this safe? Are we following an existing pattern from somewhere else in codebase?

Reply:
Safe as in, could someone structure the request to access or change things they wouldn't usually be able to. I think `cart_item_data` is usually provided by extensions? I'm not seeing any cases in core where this comes from user input.

Reply:
Yes `cart_item_data` is typically provided by extensions. It is added here so that I can supply the key/values in the link for my plugin that will then get passed to the cart [Controllers' `add_to_cart` ](https://github.com/woocommerce/woocommerce/blob/ecd7a6490b9960300e340f337c580c9715c477de/plugins/woocommerce/src/Blocks/Domain/Services/CheckoutLink.php#L190-L194)

I still think cart item data will be ignored unless a plugin is specifically configured to handle it